enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,677 @@
+pullSecret: '{"auths": ...}'
+sshKey: ssh-ed25519 AAAA...
+```
+
+The `install-config.yaml` exemplified above will generate oen MachineSet
+by subnet ID on install directory `<install-dir>/openshift/`:
+
+Example  `openshift/99_openshift-cluster-api_worker-machineset-3.yaml`:
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+metadata:
+  labels:
+    machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
+  name: ${CLUSTER_ID}-edge-us-east-1-nyc-1a  (1)
+  namespace: openshift-machine-api
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
+      machine.openshift.io/cluster-api-machineset: ${CLUSTER_ID}-edge-us-east-1-nyc-1a (1)
+  template:
+    metadata:
+      labels:
+        machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
+        machine.openshift.io/cluster-api-machine-role: edge  (2)
+        machine.openshift.io/cluster-api-machine-type: edge  (2)
+        machine.openshift.io/cluster-api-machineset: ${CLUSTER_ID}-edge-us-east-1-nyc-1a
+    spec:
+      metadata:
+        labels:
+          zone_type: local-zone (3)
+          zone_group: us-east-1-nyc-1 (4)
+          node-role.kubernetes.io/edge: "" (5)
+      taints: (5)
+        - key: node-role.kubernetes.io/edge
+          effect: NoSchedule
+      providerSpec:
+        value:
+          ami:
+            id: ${AMI_ID}
+          apiVersion: amachine.openshift.io/v1beta1
+          blockDevices:
+          - ebs: (6)
+              volumeSize: 120
+              volumeType: gp2
+          credentialsSecret:
+            name: aws-cloud-credentials
+          deviceIndex: 0
+          iamInstanceProfile:
+            id: ${CLUSTER_ID}-worker-profile
+          instanceType: c5d.2xlarge
+          kind: AWSMachineProviderConfig
+          placement:
+            availabilityZone: us-east-1-nyc-1a (7)
+            region: us-east-1
+          securityGroups:
+          - filters:
+            - name: tag:Name
+              values:
+              - ${CLUSTER_ID}-worker-sg
+          subnet:
+            id: subnetID-us-east-1-nyc-1a (8)
+          publicIp: true  (9)
+          tags:
+          - name: kubernetes.io/cluster/${CLUSTER_ID}
+            value: owned
+          userDataSecret:
+            name: worker-user-data
+```
+
+- 1) The `worker` part of the MachineSet name will be replaced with `edge`.
+- 2) MachineSet labels for the `edge` compute nodes.
+- 3) The label `zone_type` is the type of the availability zones. The [API attribute](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AvailabilityZone.html) is `ZoneType`.
+- 4) The label `zone_group` is the group to which the availability zone belongs. The [API attribute](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AvailabilityZone.html) attribute is `NetworkBorderGroup`.
+- 5) The `NoSchedule` taint will be added for each node with the
+  tag `node-role.kubernetes.io/edge=''`.
```

Comment:
Is this already recognized by Kube?

Reply:
Yes, it is. the `edge` node role will coexist with the current one `node-role.kubernetes.io/worker=''`. The idea to use that tag and label is to have a good visualization when describing the nodes:

```
 oc get nodes
NAME                                         STATUS   ROLES                  AGE   VERSION
ip-10-0-12-34.us-west-2.compute.internal     Ready    control-plane,master   32m   v1.24.0+8c7c967
ip-10-0-13-194.us-west-2.compute.internal    Ready    worker                 24m   v1.24.0+8c7c967
ip-10-0-17-18.us-west-2.compute.internal     Ready    control-plane,master   32m   v1.24.0+8c7c967
ip-10-0-17-88.us-west-2.compute.internal     Ready    worker                 14m   v1.24.0+8c7c967
ip-10-0-192-236.us-west-2.compute.internal   Ready    edge,worker            22m   v1.24.0+8c7c967
ip-10-0-194-18.us-west-2.compute.internal    Ready    edge,worker            23m   v1.24.0+8c7c967
(...)
```

Reply:
I wonder if MCO supports booting directly into custom machine config pools yet, if they do, you wouldn't have to have the worker role on these which would make things cleaner

Reply:
I am not sure, but I think if we would like to use custom MCP we may need to customize/duplicate MachineConfigs which will generate more manifests to manage and more changes on the installer, and I am not sure if could impact the lifecycle. I thought to use that one as it seems the [same approach used](https://docs.openshift.com/container-platform/4.11/machine_management/creating-infrastructure-machinesets.html#assigning-machineset-resources-to-infra-nodes) for `infra` nodes. We can use another label than `node-role.kubernetes.io/edge`, but it may be more difficult for the cluster operator/admins to identify `edge` instances on the `oc get nodes`.

Reply:
I think the approach here is reasonable so I don't see a need to change it. In terms of duplication, I believe you can extend machine configs so i don't think it would be too bad, maybe 2 or 3 additional files. But you're right, we already have precedence for this pattern with Infra so lets keep it the way it is right now

Reply:
Thanks for the feedback from you all. I created [one new section](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R228-R243) detailing what we discussed here to keep things more clear.

Reply:
> Is this already recognized by Kube?

> Yes, it is. the edge node role will coexist with the current one node-role.kubernetes.io/worker=''

This is interesting, are there any more details to elaborate on here? That is, is the node-role.kubernetes.io/edge='' specifically recognized by k8s and is there reference to this?

Reply:
> > Is this already recognized by Kube?
> 
> > Yes, it is. the edge node role will coexist with the current one node-role.kubernetes.io/worker=''
> 
> This is interesting, are there any more details to elaborate on here? That is, is the node-role.kubernetes.io/edge='' specifically recognized by k8s and is there reference to this?

I answered `recognized` as `allowed to set that label`.

According to the official documentation, only `control-plane` is a [well-known label documented on the k8s](https://kubernetes.io/docs/reference/labels-annotations-taints/). (@JoelSpeed feel free to correct me on that)

Why I proposed `node-role.kubernetes.io/edge` **to coexist** `node-role.kubernetes.io/worker`? For UX. Users can easily identify which nodes are `edge` or regular in `oc get nodes`. We could set the taints in any of the other labels [introduced in this EP](https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R241-R243), example `zone_group` without `node-role/edge`. As mentioned in my comment above, it is also a well-known approach used by `infra` "pool" in OCP, so I thought why not!

Why I didn't propose only `node-role.kubernetes.io/edge`? It will become the solution more complex, setting extra MachineConfigs, setting up kubelet, etc (I didn't map how many changes to do when choosing this path). But I thought as over scoped.